sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp

```diff
@@ -92,19 +131,176 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferAppendKernelLaunchExp(
     uint32_t numSyncPointsInWaitList,
     const ur_exp_command_buffer_sync_point_t *pSyncPointWaitList,
     ur_exp_command_buffer_sync_point_t *pSyncPoint) {
-  (void)hCommandBuffer;
-  (void)hKernel;
-  (void)workDim;
-  (void)pGlobalWorkOffset;
-  (void)pGlobalWorkSize;
-  (void)pLocalWorkSize;
-  (void)numSyncPointsInWaitList;
-  (void)pSyncPointWaitList;
-  (void)pSyncPoint;
+  // Preconditions
+  UR_ASSERT(hCommandBuffer->Context == hKernel->getContext(),
+            UR_RESULT_ERROR_INVALID_KERNEL);
+  UR_ASSERT(workDim > 0, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
+  UR_ASSERT(workDim < 4, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
+
+  ur_result_t Result = UR_RESULT_SUCCESS;
+  CUgraphNode GraphNode;
+
+  std::vector<CUgraphNode> DepsList;
+  UR_CALL(getNodesFromSyncPoints(hCommandBuffer, numSyncPointsInWaitList,
+                                 pSyncPointWaitList, DepsList));
+
+  if (*pGlobalWorkSize == 0) {
+    // Create a empty node is the kernel worload size is zero
+    Result = UR_CHECK_ERROR(
+        cuGraphAddEmptyNode(&GraphNode, hCommandBuffer->MCudaGraph,
+                            DepsList.data(), DepsList.size()));
+
+    // Get sync point and register the event with it.
+    *pSyncPoint = hCommandBuffer->GetNextSyncPoint();
+    hCommandBuffer->RegisterSyncPoint(*pSyncPoint,
+                                      std::make_shared<CUgraphNode>(GraphNode));
+    return Result;
+  }
 
-  detail::ur::die("Experimental Command-buffer feature is not "
-                  "implemented for CUDA adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  // Set the number of threads per block to the number of threads per warp
```

Comment:
Quite a lot of this code looks like it's copied from the implementation of `urEnqueueKernelLaunch`, which makes it harder to maintain. Can we refactor the implementation of `urEnqueueKernelLaunch` into some helper functions that be reused in `urCommandBufferAppendKernelLaunchExp`?

Reply:
I refactored the code and created a helper used by urEnqueueKernelLaunch and urCommandBufferAppendKernelLaunchExp.